### PR TITLE
Sped up FlutterStandardCodec writing speed.

### PR DESCRIPTION
**Pull request by gaaclarke**
This improves encoding time 18% (based on the platform channel large benchmark).

This follows up on the investigation done in https://github.com/flutter/engine/pull/37883.  After this lands we should be almost back where we were before the ARC migration.

This includes the following optimizations:
1) A big win was made from avoiding `[NSNull null]` since it caused alot of ARC activity
1) Iterating over maps doesn't require looking up the values now
1) NSStrings in utf8 format will get read from directly (I'm not sure how this compares with objc logic)
1) For common datatypes no arc should be used (e.g. a List with Strings and Ints)
1) For common datatypes recursion happens with c function invocations instead of objc_msgSend

There is no functional change here, just performance change.

before:
Screen Shot 2022-12-15 at 11 04 36 PM

after:
Screen Shot 2022-12-15 at 10 58 59 PM

There are existing tests that cover the PR:
 - platform channel benchmarks
 - unit tests in the engine
 - integration tests in the framework

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/darwin/common/framework/Source/FlutterStandardCodecHelper.h [hunk 1]
```diff
@@ -62,6 +73,27 @@ CFTypeRef FlutterStandardCodecHelperReadValueOfType(
     CFTypeRef (*ReadTypedDataOfType)(FlutterStandardField, CFTypeRef),
     CFTypeRef user_data);
 
+//////////////////////////////////////////////
```
Comment (cyanglaz) [inline shell/platform/darwin/common/framework/Source/FlutterStandardCodecHelper.h:76]: What is this line for?

Comment (gaaclarke) [reply shell/platform/darwin/common/framework/Source/FlutterStandardCodecHelper.h:76]: Just a split between the reader functions and the writer functions.

Comment (cyanglaz) [reply shell/platform/darwin/common/framework/Source/FlutterStandardCodecHelper.h:76]: Maybe Something more descriptive? I'm not sure if there are any c++ header standard separator like #pragma in Objective C.

Comment (gaaclarke) [reply shell/platform/darwin/common/framework/Source/FlutterStandardCodecHelper.h]: I added comments that will group them together in our doxygen output.

**Review by cyanglaz — APPROVED** [review]
LGTM

**Comment by Hixie** [comment]
test-exempt: code refactor with no semantic change

**Comment by Hixie** [comment]
Nice performance boost, too!

**Pull request merged**